pkg/controllers/management/k3supgrade/register.go

```diff
@@ -0,0 +1,209 @@
+	// TODO: We shouldn't create one every time
+	planConfig, err := planClientset.NewForConfig(&clusterCtx.RESTConfig)
+	if err != nil {
+		return err
+	}
+	planClient := planConfig.Plans(metav1.NamespaceAll)
+
+	planList, err := planClient.List(metav1.ListOptions{})
+	if err != nil {
+		// may need to handle this error, if there is no Plan CRD what should we do?
+		if errors.IsNotFound(err) {
+			// no plan CRD exists
+			logrus.Warnf("plan CRD does not exist: %s", err)
+		}
+		return err
+	}
+	masterPlan := planv1.Plan{}
+	workerPlan := planv1.Plan{}
+	// deactivate all existing plans that are not managed by Rancher
+	for _, plan := range planList.Items {
+		if _, ok := plan.Labels[rancherManagedPlan]; !ok {
+			// inverse selection is used here, we select a non-existent label
+			plan.Spec.NodeSelector = &metav1.LabelSelector{
+				MatchExpressions: []metav1.LabelSelectorRequirement{{
+					Key:      upgradeDisableLabelKey,
+					Operator: metav1.LabelSelectorOpExists,
+				}}}
+
+			_, err = planClient.Update(&plan)
+			if err != nil {
+				return err
+			}
+		} else {
+			// if any of the rancher plans are currently applying, set updating status on cluster
+			if len(plan.Status.Applying) > 0 {
+				v3.ClusterConditionUpdated.Unknown(cluster)
+				cluster, err = h.clusterClient.Update(cluster)
+				if err != nil {
+					return err
+				}
+			} else {
+				//set it back if not
+				if v3.ClusterConditionUpdated.IsUnknown(cluster) {
+					v3.ClusterConditionUpdated.True(cluster)
+					cluster, err = h.clusterClient.Update(cluster)
+					if err != nil {
+						return err
+					}
+				}
+			}
+
+			switch name := plan.Name; name {
+			case k3sMasterPlanName:
+				masterPlan = plan
+			case k3sWorkerPlanName:
+				workerPlan = plan
+			}
+		}
+	}
+
+	// if rancher plans exist, do we need to update?
+	if masterPlan.Name != "" || workerPlan.Name != "" {
+		if masterPlan.Name != "" {
+			newMaster, err := configureMasterPlan(masterPlan, cluster.Spec.K3sConfig.Version, cluster.Spec.K3sConfig.ServerConcurrency)
+			if err != nil {
+				return err
+			}
+			if !cmp(masterPlan, newMaster) {
+				planClient = planConfig.Plans(systemUpgradeNS)
+				_, err = planClient.Update(&newMaster)
+				if err != nil {
+					return err
+				}
+			} else {
+				fmt.Println("master plan is the same, not updating")
+			}
+		}
+
+		if workerPlan.Name != "" {
+			newWorker, err := configureWorkerPlan(workerPlan, cluster.Spec.K3sConfig.Version, cluster.Spec.K3sConfig.WorkerConcurrency)
+			if err != nil {
+				return err
+			}
+			if !cmp(workerPlan, newWorker) {
+				planClient = planConfig.Plans(systemUpgradeNS)
+				_, err = planClient.Update(&newWorker)
+				if err != nil {
+					return nil
+				}
+			} else {
+				fmt.Println("worker plan is the same, not updating")
+			}
+		}
+
+	} else { // create the plans
+		planClient = planConfig.Plans(systemUpgradeNS)
+		masterPlan, err = generateMasterPlan(cluster.Spec.K3sConfig.Version,
```

Comment:
do we not want to error check this?

Reply:
dropped error creation in `generateMasterPlan`